[JENKINS-49196] java.io.IOException: Premature EOF when shutdown CLI command is used

MRamonLeon opened the pull request:
Add a new thread to exit Jenkins for the shutdown cli command avoiding the Premature EOF IOException.

See [JENKINS-49196](https://issues.jenkins-ci.org/browse/JENKINS-49196).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* [JENKINS-49196](https://issues.jenkins-ci.org/browse/JENKINS-49196): avoid _Premature EOF_ error when using the `shutdown` CLI command .

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [X] Manually tested with ssh, remoting and https
- [ ] N/A For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees @dwnusbaum 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

MRamonLeon committed:
[JENKINS-49196] Shutdown in a new thread to avoid PrematureEOF

Add a new thread to exit Jenkins for the shutdown cli command to avoid
the Premature EOF IOException

Wadeck reviewed (approved):
The code seems good, did you try to write a test for that particular case?

fcojfernandez reviewed (commented):
Code LGTM, but there are conflicts that have to be fixed first. Pending the fixes to approve the PR

dwnusbaum commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -4266,18 +4266,29 @@ protected RestartCause() {
+                    ACL.impersonate(ACL.SYSTEM);
+                    LOGGER.severe(String.format("Shutting down VM as requested by %s from %s",
+                            getAuthentication().getName(), req!=null?req.getRemoteAddr():"???"));
+                    if (rsp!=null) {
```
I think all of the stuff that modifies the response needs to happen outside of the thread after `start` is called. Otherwise, the method will have already returned, so I'm not sure what would even happen if you modify the response.

MRamonLeon committed:
Merge with master

MRamonLeon committed:
[JENKINS-49196] Shutdown in a new thread to avoid PrematureEOF

Add a new thread to exit Jenkins for the shutdown cli command to avoid
the Premature EOF IOException

MRamonLeon commented:
@dwnusbaum I've changed the location of the *rsp* stuff, same behavior calling from terminal. But if it's called from a browser then nothing is shown if the stuff is in the Thread, whereas the message is shown when it's in the main thread.

dwnusbaum reviewed (approved):
Looks fine to me. I guess you can't add an automated test using CLICommandInvoker because of the call to `System.exit`? Maybe you could work around it with a custom security manager or something?

CC @oleg-nenashev since I know he had an idea for a different way to approach the issue in case he wants to take a look.

varyvol commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -4283,21 +4283,31 @@ protected RestartCause() {
+            }
+        }.start();
+
         if (rsp!=null) {
```
Shouldn't this be done before calling start?

MRamonLeon replied:
It's not necessary because the cleanUp lasts a while, but I've changed it just in case

varyvol commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -4283,21 +4283,31 @@ protected RestartCause() {
+            @Override
+            public void run() {
+                try {
+                    ACL.impersonate(ACL.SYSTEM);
```
Why is the impersonation necessary?

MRamonLeon commented:
@dwnusbaum I've tried to use a `SecurityManager` as I used in the file-leak-detector-plugin proposal catching the attempt to exit but it seems there is a conflict with some _SecurityContext_ set in the threads during the cleanUp. It gives me weird exceptions when the code gets the logger (for example):

```
8.508 [id=21]	SEVERE	hudson.triggers.SafeTimerTask#run: Timer task hudson.TcpSlaveAgentListener$TcpSlaveAgentListenerRescheduler@24dbce15 failed
java.security.AccessControlException: access denied ("java.util.PropertyPermission" "sun.util.logging.disableCallerCheck" "read")
```

I wont continue investing in these tests.

MRamonLeon committed:
[JENKINS-49196] Response first just in case the thread ends before

It's not necessary because the clean up last a while, but just in case

MRamonLeon commented:
Closing so the pr-builder can be re-launched

batmat commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -4293,10 +4291,22 @@ public void doExit( StaplerRequest req, StaplerResponse rsp ) throws IOException
+                    cleanUp();
+                    System.exit(0);
+                } catch (Exception e) {
+                    LOGGER.log(Level.WARNING, "Failed to shut down Jenkins", e);
```
Probably should be an error, it looks like it's more than just a warning. No big deal though.

batmat commented:
I am going to merge this later today if nobody objects or does it in the meantime, given the number of existing reviews & approvals.

batmat commented:
Adjusted the changelog a bit for hopefully a bit more end-user understandability, also marked the related JIRA `lts-candidate`. Merging.
